### PR TITLE
Optimize CcInfo computation

### PR DESCRIPTION
**Pull request by dzbarsky**
This shows up around 4% of buildbuddy analysis phase
image

After this change, it goes away.

It seems like Bazel is able to make this lazy, which is good since it's only used for building LINKMODE_C_ARCHIVE/LINKMODE_C_SHARED which is fairly rare.

**What type of PR is this?**
Performance improvement

**What does this PR do? Why is it needed?**
Avoid computing CcInfo for every intermediate library when it's not needed

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

File: go/private/rules/binary.bzl [hunk 1]
```diff
@@ -94,6 +95,24 @@ def new_cc_import(
         ),
     )
 
+def _go_cc_aspect_impl(target, _ctx):
```
Comment (fmeum) [inline go/private/rules/binary.bzl:98]: Why `_ctx`? It's not unused.

**Review by fmeum — COMMENTED** [review]
I'm surprised to see that this reduces CPU usage as I would have expected the aspect to be evaluated eagerly. Could you check the memory implications of having a new aspect?

File: go/private/rules/binary.bzl [hunk 2]
```diff
@@ -94,6 +95,24 @@ def new_cc_import(
...
+    if GoSource in target:
```
Comment (fmeum) [inline go/private/rules/binary.bzl:99]: Let's make this an early return instead.

File: go/private/rules/binary.bzl (hunk 1 above)
Comment (dzbarsky) [reply go/private/rules/binary.bzl:98]: oh it was leftover from when i was using the GoSource, i'll fix

**Comment by dzbarsky** [comment]
> I'm surprised to see that this reduces CPU usage as I would have expected the aspect to be evaluated eagerly. Could you check the memory implications of having a new aspect?

I did as well, but I guess it must be lazy somehow? (And if it's not, I think we have a path to make it so; the `go_binary` macro already splits to 2 rules for executable vs non-executable, and we only need the aspect on the non-executable one)

I'll poke at memory usage as well

**Comment by dzbarsky** [comment]
> > I'm surprised to see that this reduces CPU usage as I would have expected the aspect to be evaluated eagerly. Could you check the memory implications of having a new aspect?
> 
> I did as well, but I guess it must be lazy somehow? (And if it's not, I think we have a path to make it so; the `go_binary` macro already splits to 2 rules for executable vs non-executable, and we only need the aspect on the non-executable one)
> 
> I'll poke at memory usage as well

@fmeum so I tried `bazel shutdown && bazel build --nobuild //... && bazel info used-heap-size-after-gc` on buildbuddy repo. I saw 215-216MB as baseline, and 220MB with this PR. I made the aspect only apply to non-executable binary and now it is down to 213-214MB, so it's actually an improvement :)

**Review by fmeum — APPROVED** [review]
Great job!

**Pull request merged**